Write Docker guide (#21)
MarsBased/handbook: guides/development/docker-guide.md

If gems were installed in the image then when there is a change in the used gems (like adding a new gem or updating one), the image needs to be rebuilt, thus needing to install **all gems** again (which can take a long time). By, instead, having gems installed on a volume, when there is a change in a gem, we can just run `bundle install` again (like we would do locally) and it will just install that new gem or updated version.

It is useful to think of it as if the Docker image is just the bare-bones OS and you do the rest the same way you would do it locally (`bundle install`, `npm install`, etc.). More specifically, the image only contains the programming language and system packages (like ImageMagick or FFmpeg).

## Rails Docker for development setup

In order to Dockerize a Rails application for development, we create a `.devcontainer` directory in the application root, which contains the `Dockerfile`, `docker-compose.yml` and other support files. Separating it into its own directory avoids mixing it with the production Docker setup which usually resides in the root.

> When copying the files from this repo you need to replace several values for the appropriate in your application. These values are: `<application-name>`, `<postgres-version>`, `<redis-version>`, `ruby-version>`, `<major-node-version>` and `<bundler-version>`.

Author: Thanks!

It is useful to think of it as if the Docker image is just the bare-bones OS and you do the rest the same way you would do it locally (`bundle install`, `npm install`, etc.). More specifically, the image only contains the programming language and system packages (like ImageMagick).

There are 2 options to setup the development environment with Docker:
* Services only: External services (database, Redis, Minio, Elastic Search, etc.) are run with Docker but the application runs locally.
* Services + Application: Apart from services, the application is also run in a container.

Running everything with Docker has the advantage of not needing to have any dependency installed locally, apart from Docker. The disadvantage is that it runs slower because the application can't use the full memory + CPU potential from the computer.

> When working on a single application it makes sense to use the services only approach, while when working on multiple projects it is more convenient to run all applications in containers to avoid having a dependency hell in the computer.

Reviewer: "Application" word is confusing (to me). 

```suggestion
When working on a single project, it makes sense to use the services only approach, while when working on several projects it is more convenient to run each inside containers to avoid having a dependency hell in the computer.
```
